### PR TITLE
Decouple calculate fee from Wallet

### LINKED ISSUE #2352: [Question] Does CalculateNetworkFee(DataCache snapshot, Transaction tx) needs to be inside Wallet? Make network fee calculation more generic.
The call https://github.com/neo-project/neo/blob/d73d190c87eb999c36934c1fe93aaee8c818701a/src/neo/Wallets/Wallet.cs#L356-L421 is currently inside Wallet and this makes it requires a Dummy Wallet on neo-modules RpcServer https://github.com/neo-project/neo-modules/blob/ccc2ff93599c8127ad2d5a2a21f99360d84c79b3/src/RpcServer/RpcServer.Wallet.cs#L124.

What necessarily makes it need an open wallet to make this calculation?

On NEO 2 perhaps yes, due to UTXO calculations regarding size.
However, now, we mostly just use `UInt160[] hashes = tx.GetScriptHashesForVerifying(snapshot);` which are already created and then we get the `VerificationScript` using `GetAccount(hash)?.Contract?.Script;`

I suggested to restructure these functions to a different place on the neo-core and even move them from the Rpc-Wallet, making them more user friendly in a way that `VerificationScripts` can be passed as parameters for a generic calculation.

### PR DESCRIPTION
**Pull request by shargon**
Copy of https://github.com/neo-project/neo/pull/2354
Close #2352

### CODE REVIEW

File: src/Neo/Wallets/Helper.cs [hunk 1]
```diff
@@ -72,5 +77,91 @@ internal static byte[] XOR(byte[] x, byte[] y)
...
+        public static long CalculateNetworkFee(this Transaction tx, DataCache snapshot, ProtocolSettings settings, Func<UInt160, byte[]> accountScript, long maxExecutionCost = ApplicationEngine.TestModeGas)
+        {
+            UInt160[] hashes = tx.GetScriptHashesForVerifying(snapshot);
+
+            // base size for transaction: includes const_header + signers + attributes + script + hashes
+            int size = Transaction.HeaderSize + tx.Signers.GetVarSize() + tx.Attributes.GetVarSize() + tx.Script.GetVarSize() + IO.Helper.GetVarSize(hashes.Length);
+            uint exec_fee_factor = NativeContract.Policy.GetExecFeeFactor(snapshot);
+            long networkFee = 0;
+            int index = -1;
+            foreach (UInt160 hash in hashes)
```
Comment (shargon) [inline src/Neo/Wallets/Helper.cs:99]:
```suggestion
            int size = Transaction.HeaderSize + tx.Signers.GetVarSize() + tx.Attributes.GetVarSize() + tx.Script.GetVarSize() + IO.Helper.GetVarSize(hashes.Length), index = -1;
            uint exec_fee_factor = NativeContract.Policy.GetExecFeeFactor(snapshot);
            long networkFee = 0;
            foreach (UInt160 hash in hashes)
```

**Comment by shargon** [comment]
@Jim8y merge?

**Comment by vncoelho** [comment]
I did not remember that I created this issue...aheuaheuaea
Let me review

File: src/Neo/Wallets/Wallet.cs [hunk 2]
```diff
@@ -578,99 +578,12 @@ private Transaction MakeTransaction(DataCache snapshot, ReadOnlyMemory<byte> scr
                     tx.SystemFee = engine.GasConsumed;
                 }
 
-                tx.NetworkFee = CalculateNetworkFee(snapshot, tx, maxGas);
+                tx.NetworkFee = tx.CalculateNetworkFee(snapshot, ProtocolSettings, (a) => GetAccount(a)?.Contract?.Script, maxGas);
```
Comment (vncoelho) [inline src/Neo/Wallets/Wallet.cs:581]: isn´t this function a little complex `ProtocolSettings, (a) => GetAccount(a)?.Contract?.Script`?
Maybe we should just pass the Script

Comment (shargon) [reply src/Neo/Wallets/Wallet.cs:581]: There can be a lot of scripts

**Review by vncoelho — APPROVED** [review]

**Comment by dusmart** [comment]
It's an important public function maybe used by some plugins or elsewhere. The best practice is to keep that function and mark it obsolete instead of deleting it directly.

Although this has been merged, I have to reminder that NEO N3 is a sophisticated blockchain now and we'd better providing compatibility to others as much as possible.

**Pull request merged**